pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FunnelCountAggregationFunction.java

```diff
@@ -35,46 +50,107 @@
+ *  <p><ul>
+ *  <li>'set': See DISTINCTCOUNT at {@link DistinctCountAggregationFunction}
+ *  <li>'bitmap' (default): See DISTINCTCOUNTBITMAP at {@link DistinctCountBitmapAggregationFunction}
+ *  <li>'theta_sketch': See DISTINCTCOUNTTHETASKETCH at {@link DistinctCountThetaSketchAggregationFunction}
+ *  <li>'partitioned': See SEGMENTPARTITIONEDDISTINCTCOUNT {@link SegmentPartitionedDistinctCountAggregationFunction}
+ *  <li>'sorted': sorted counts per segment then sums up. Only availabe in combination with 'partitioned'.
+ *  <li>'nominalEntries=4096': theta sketch configuration, default is 4096.
+ *  </ul><p>
  */
-public class FunnelCountAggregationFunction implements AggregationFunction<List<Long>, LongArrayList> {
+public class FunnelCountAggregationFunction implements AggregationFunction<Object, LongArrayList> {
+  private static final Sketch EMPTY_SKETCH = new UpdateSketchBuilder().build().compact();
   final List<ExpressionContext> _expressions;
   final List<ExpressionContext> _stepExpressions;
   final List<ExpressionContext> _correlateByExpressions;
   final ExpressionContext _primaryCorrelationCol;
   final int _numSteps;
-
-  final SegmentAggregationStrategy<?, List<Long>> _sortedAggregationStrategy;
-  final SegmentAggregationStrategy<?, List<Long>> _bitmapAggregationStrategy;
+  final int _nominalEntries;
+  final boolean _partitionSetting;
+  final boolean _sortingSetting;
+  final boolean _thetaSketchSetting;
+  final boolean _setSetting;
+
+  final AggregationStrategy _thetaSketchAggregationStrategy;
+  final AggregationStrategy _bitmapAggregationStrategy;
+  final AggregationStrategy _sortedAggregationStrategy;
+
+  final MergeStrategy _thetaSketchMergeStrategy;
+  final MergeStrategy _setMergeStrategy;
+  final MergeStrategy _bitmapMergeStrategy;
+  final MergeStrategy _partitionedMergeStrategy;
+
+  final ResultExtractionStrategy _thetaSketchResultExtractionStrategy;
+  final ResultExtractionStrategy _setResultExtractionStrategy;
+  final ResultExtractionStrategy _bitmapResultExtractionStrategy;
+  final ResultExtractionStrategy _sortedPartitionedResultExtractionStrategy;
+  final ResultExtractionStrategy _bitmapPartitionedResultExtractionStrategy;
+  final ResultExtractionStrategy _thetaSketchPartitionedResultExtractionStrategy;
 
   public FunnelCountAggregationFunction(List<ExpressionContext> expressions) {
     _expressions = expressions;
+    Option.validate(expressions);
     _correlateByExpressions = Option.CORRELATE_BY.getInputExpressions(expressions);
     _primaryCorrelationCol = Option.CORRELATE_BY.getFirstInputExpression(expressions);
     _stepExpressions = Option.STEPS.getInputExpressions(expressions);
     _numSteps = _stepExpressions.size();
-    _sortedAggregationStrategy = new SortedAggregationStrategy();
+
+    final List<String> settings = Option.SETTINGS.getLiterals(expressions);
+    Setting.validate(settings);
```

Comment:
Should we validate only one of the strategies are selected and return an error otherwise?

Reply:
Some combinations are valid. For example, one could ask for `partitioned` together with `sorted` or together with `tetha_sketch`.
But we could indeed check for invalid combinations instead of just prioritising them.